### PR TITLE
fix(manager-server): sanitize usage credentials before persistence

### PR DESCRIPTION
**Pull request by sjcoder**
## Summary

Add a final credential-redaction boundary immediately before usage events are written to SQLite. This closes paths where producer-built events could bypass the existing raw-payload normalizer and persist secrets in `source`, structured metadata, valid `raw_json`, or `fail_body`.

## Scope

- [ ] Frontend panel
- [x] Manager Server
- [x] CPA panel mode
- [x] Full Docker mode
- [ ] Native packages / release
- [ ] Docs / Wiki
- [ ] CI / build / tooling

## Changes

- Sanitize every persisted usage-event string field at `InsertBatch`, regardless of which producer built the event.
- Recursively redact token-shaped values in valid JSON and response metadata, including structured management-key fields and CPAMP key shapes.
- Preserve complete sanitized local failure bodies while keeping public failure summaries bounded.
- Hash credential-bearing identity fields instead of collapsing them to a shared redaction marker.
- Add synthetic unit and SQLite round-trip regressions while preserving non-secret filenames and existing account identity behavior.

## User Impact

Usage monitoring continues to work normally, but raw credentials are removed before SQLite persistence.

## Compatibility / Runtime Notes

- CPA panel mode: The shared Manager Server persistence boundary is covered.
- Manager Server mode: No schema or configuration change.
- Full Docker / native packages: No packaging change; behavior activates with the updated Manager Server binary.

## Data / Security Notes

This is a defense-in-depth storage boundary for usage data. Tests use synthetic markers only. The change does not migrate historical rows; operators should clean existing affected records and rotate exposed credentials after deploying a fixed release.

## Risk / Rollback

Risk level: Medium

Rollback notes: Revert these commits. The main compatibility risk is over-redaction, covered by the full Manager Server suite plus explicit filename, account-identity, long-diagnostic, structured-key, and every-string-field regressions.

## Verification

- [ ] Type check
- [ ] Lint
- [x] Tests
- [ ] Build
- [ ] Manual UI check
- [ ] Docs/link check
- [ ] Not applicable, docs-only

Commands / evidence:

```text
cd apps/manager-server
go test ./internal/usage ./internal/repository/usageevent -count=1
# PASS

go test ./...
# PASS
```

## Screenshots / Recordings

N/A — backend-only behavior.

## Docs

- [x] Not needed — explanation included below

Docs decision: No user-facing capability or configuration changed.

### CODE REVIEW

File: apps/manager-server/internal/usage/event.go [hunk 1]
```diff
@@ -1070,6 +1092,8 @@ func redactValueWithParent(parentKey string, value any) any {
 			result = append(result, redactValueWithParent(parentKey, child))
 		}
 		return result
+	case string:
+		return redactCredentialText(item)
```
[the remark this answers is not in the document]
Comment (sjcoder) [reply apps/manager-server/internal/usage/event.go:1148]: Fixed in d9326d4. Structured secret-key detection now includes normalized `management_key` and `managementkey`, with a regression using an ordinary unprefixed camelCase `managementKey` value.

File: apps/manager-server/internal/usage/event.go [hunk 2]
```diff
@@ -995,15 +996,36 @@ func FailSummaryFromBody(body string) string {
...
+// SanitizeForPersistence applies the final credential-redaction boundary to
+// fields that may contain untrusted upstream payloads. Callers can normalize
+// earlier for display and aggregation, but no event should reach storage
+// without passing through this function.
+func SanitizeForPersistence(event Event) Event {
+	// Preserve existing source identity semantics (including account emails)
+	// while removing credentials that slipped past producer normalization.
+	event.Source = redactCredentialText(event.Source)
+	event.FailSummary = FailSummaryFromBody(event.FailSummary)
+	event.FailBody = FailSummaryFromBody(event.FailBody)
```
[the remark this answers is not in the document]
Comment (sjcoder) [reply apps/manager-server/internal/usage/event.go]: Fixed in d9326d4. `fail_body` now uses credential redaction without the summary length cap; `fail_summary` remains bounded. The regression verifies a body longer than 4,096 bytes retains its trailing diagnostic text after the token is removed, and the full Manager Server suite passes.

**Comment by sjcoder** [comment]
@codex review

**Comment by sjcoder** [comment]
@codex review

File: apps/manager-server/internal/repository/usageevent/repository.go [hunk 3]
```diff
@@ -130,6 +130,7 @@ func (r *repository) InsertBatch(ctx context.Context, events []model.UsageEvent)
 
 	result := model.InsertResult{}
 	for _, event := range events {
+		event = usage.SanitizeForPersistence(event)
```
[the remark this answers is not in the document]
Comment (sjcoder) [reply apps/manager-server/internal/repository/usageevent/repository.go]: Fixed in 595f29a. Only the cheap credential-safe event-hash transform runs before the identity-ledger claim; full field/JSON sanitization now runs after duplicates are skipped. Insert reuses the sanitized `RawJSON` instead of calling `SafeRawJSON` again. The full Manager Server suite passes.

File: apps/manager-server/internal/usage/event.go [hunk 4]
```diff
@@ -439,9 +439,10 @@ var (
 	endpointPattern          = regexp.MustCompile(`^(GET|POST|PUT|PATCH|DELETE|OPTIONS|HEAD)\s+(\S+)`)
 	authorizationHeaderRegex = regexp.MustCompile(`(?i)\b(authorization\s*[:=]\s*)(?:bearer\s+)?[^\s,"'{}]+`)
 	bearerTokenRegex         = regexp.MustCompile(`(?i)\bbearer\s+[A-Za-z0-9._~+/=-]{8,}`)
-	apiKeyTokenRegex         = regexp.MustCompile(`(sk-proj-[A-Za-z0-9-_]{6,}|sk-ant-[A-Za-z0-9-_]{6,}|sk-[A-Za-z0-9-_]{6,}|sess-[A-Za-z0-9-_]{6,}|ghp_[A-Za-z0-9]{6,}|github_pat_[A-Za-z0-9_]{20,}|AIza[0-9A-Za-z-_]{8,}|hf_[A-Za-z0-9]{6,}|pk_[A-Za-z0-9]{6,}|rk_[A-Za-z0-9]{6,})`)
+	apiKeyTokenRegex         = regexp.MustCompile(`(sk-proj-[A-Za-z0-9-_]{6,}|sk-ant-[A-Za-z0-9-_]{6,}|sk-[A-Za-z0-9-_]{6,}|sess-[A-Za-z0-9-_]{6,}|ghp_[A-Za-z0-9]{6,}|github_pat_[A-Za-z0-9_]{20,}|AIza[0-9A-Za-z-_]{8,}|hf_[A-Za-z0-9]{6,}|pk_[A-Za-z0-9]{6,}|rk_[A-Za-z0-9]{6,}|cpamp_[A-Za-z0-9-_]{6,})`)
```
[the remark this answers is not in the document]
Comment (sjcoder) [reply apps/manager-server/internal/usage/event.go]: Fixed in 595f29a. The CPAMP admin-key alternative now matches the generator contract exactly: `cpamp_` plus 32 alphanumeric characters and a word boundary. A regression verifies `cpamp_account.json` remains unchanged alongside the existing ordinary filename case.

**Comment by sjcoder** [comment]
@codex review

File: apps/manager-server/internal/usage/event.go [hunk 5]
```diff
@@ -1092,6 +1168,8 @@ func isSecretKey(key string) bool {
 	normalized := normalizeSecretKey(key)
 	return normalized == "api_key" ||
 		normalized == "apikey" ||
+		normalized == "management_key" ||
+		normalized == "managementkey" ||
```
[the remark this answers is not in the document]
Comment (sjcoder) [reply apps/manager-server/internal/usage/event.go:1174]: Fixed in 939b978. Structured cpaManagementKey, cpa-management-key, x-api-key, and xApiKey aliases are now treated as secret keys, with regression coverage.

File: apps/manager-server/internal/usage/event.go [hunk 6]
```diff
@@ -439,9 +439,10 @@ var (
 	endpointPattern          = regexp.MustCompile(`^(GET|POST|PUT|PATCH|DELETE|OPTIONS|HEAD)\s+(\S+)`)
 	authorizationHeaderRegex = regexp.MustCompile(`(?i)\b(authorization\s*[:=]\s*)(?:bearer\s+)?[^\s,"'{}]+`)
 	bearerTokenRegex         = regexp.MustCompile(`(?i)\bbearer\s+[A-Za-z0-9._~+/=-]{8,}`)
-	apiKeyTokenRegex         = regexp.MustCompile(`(sk-proj-[A-Za-z0-9-_]{6,}|sk-ant-[A-Za-z0-9-_]{6,}|sk-[A-Za-z0-9-_]{6,}|sess-[A-Za-z0-9-_]{6,}|ghp_[A-Za-z0-9]{6,}|github_pat_[A-Za-z0-9_]{20,}|AIza[0-9A-Za-z-_]{8,}|hf_[A-Za-z0-9]{6,}|pk_[A-Za-z0-9]{6,}|rk_[A-Za-z0-9]{6,})`)
+	apiKeyTokenRegex         = regexp.MustCompile(`(sk-proj-[A-Za-z0-9-_]{6,}|sk-ant-[A-Za-z0-9-_]{6,}|sk-[A-Za-z0-9-_]{6,}|sess-[A-Za-z0-9-_]{6,}|ghp_[A-Za-z0-9]{6,}|github_pat_[A-Za-z0-9_]{20,}|AIza[0-9A-Za-z-_]{8,}|hf_[A-Za-z0-9]{6,}|pk_[A-Za-z0-9]{6,}|rk_[A-Za-z0-9]{6,}|cpamp_[A-Za-z0-9]{32}\b)`)
```
[the remark this answers is not in the document]
Comment (sjcoder) [reply apps/manager-server/internal/usage/event.go]: Fixed in 939b978. The short token prefixes now require at least 20 credential characters, and regression tests preserve the reported hf_account.json case plus matching ghp, pk, rk, and sess filenames while retaining positive redaction coverage.

**Comment by sjcoder** [comment]
@codex review

**Comment by seakee** [comment]
Thanks for the work on this PR. The persistence-boundary sanitization direction is valuable, and the review helped uncover several deeper issues around usage-event identity, deduplication, historical compatibility, and SQLite write-path behavior.

At this point, the remaining changes are no longer limited to credential redaction and involve core CPAMP data-contract decisions. We’re going to take over the implementation on the maintainer side rather than ask you to continue iterating on these broader architectural concerns.

We’ll preserve attribution for any code, tests, or implementation work from this PR that is materially reused in the replacement fix.

Thanks again for identifying and working on this issue.

**Comment by seakee** [comment]
Thanks for the original fix and for identifying the need for a final credential-sanitization boundary before UsageEvent persistence.

This work has now been superseded by #749, which was merged into `dev` and carries the final implementation after additional review and hardening. The merged version keeps the same persistence-boundary direction while extending coverage for structured/malformed payloads, Authorization/Cookie cases, hash handling, import validation, regression tests, and related edge cases.

Your contribution is preserved in the merged history via `Co-authored-by: Daniel Alvarez <daniel@sjcoder.com>`.

Closing this PR as superseded by #749. Thanks again for the contribution.
